raft/src/main/java/org/apache/kafka/snapshot/Snapshots.java

```diff
@@ -92,4 +100,22 @@ public static Path createTempFile(Path logDir, OffsetAndEpoch snapshotId) throws
 
         return Optional.of(new SnapshotPath(path, new OffsetAndEpoch(endOffset, epoch), partial));
     }
+
+    /**
+     * Delete this snapshot from the filesystem.
+     */
+    public static CompletableFuture<Boolean> deleteSnapshotIfExists(Path logDir, OffsetAndEpoch snapshotId) throws IOException {
+        Path path = snapshotPath(logDir, snapshotId);
+        Path destination = Snapshots.deleteRename(path, snapshotId);
+        Utils.atomicMoveWithFallback(path, destination);
+
+        return CompletableFuture.supplyAsync(() -> {
```

Comment:
This looks like the Jira and commit that introduced a delay when deleting segments: https://issues.apache.org/jira/browse/KAFKA-636. It looks like the delays is there to reduce the probability that the file is deleted while Kafka is still reading the file.

Even though `CompletableFuture::supplyAsync` run async it will attempt to run immediately.

Reply:
If the file is deleted while Kafka is still reading the file, won't the reading fail since we have already renamed the log file?

Reply:
Probably not since the reference would be through the file descriptor.